### PR TITLE
Write through support

### PR DESCRIPTION
**Pull request by MeirShpilraien**

### CODE REVIEW

File: src/main/java/com/redislabs/Source.java [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+                  } catch (Exception e) {
+                    GearsBuilder.log(e.toString(), LogLevel.WARNING);
+                    return null;
+                  }
+                }
+              }
+              return f;
+            }
+      };
+
```
Comment (MeirShpilraien) [inline src/main/java/com/redislabs/Source.java:183]: Put all those in a function

File: src/main/java/com/redislabs/Source.java [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+                }
+              }
+              return f;
+            }
+      };
+      
+      CommandOverrider hsetco = new CommandOverrider().setCommand("hset");
+      GearsBuilder.CreateGearsBuilder(hsetco, "hset command override").
+      asyncMap(asyncMapper).register(ExecutionMode.ASYNC_LOCAL, 
+          ()->{writeThroughIsRegistered=true;}, ()->{writeThroughIsRegistered=false;});
```
Comment (MeirShpilraien) [inline src/main/java/com/redislabs/Source.java:187]: Is this correct?

**Comment by MeirShpilraien** [comment]
- [ ] Handle remove source
- [ ] Handle remove connector

**Review by gkorland — CHANGES_REQUESTED** [review]

File: src/test/resources/mysql_hibernate.cfg.xml [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+        "-//Hibernate/Hibernate Configuration DTD 3.0//EN"
+	"http://www.hibernate.org/dtd/hibernate-configuration-3.0.dtd">
+        
+<hibernate-configuration>
+	<session-factory>
+		<!-- JDBC Database connection settings -->
+		<property name="connection.driver_class">org.mariadb.jdbc.Driver</property>
+		<property name="connection.url">jdbc:mysql://localhost:3306/test?allowPublicKeyRetrieval=true&amp;useSSL=false</property>
+		<!--property name="hibernate.connection.driver_class">oracle.jdbc.driver.OracleDriver</property>
+		<property name="hibernate.connection.url">jdbc:oracle:thin:@localhost:1521:xe</property>
```
Comment (gkorland) [inline src/test/resources/mysql_hibernate.cfg.xml:11]: @MeirShpilraien I think you should remove all the commented configuration to make it simple and clean

File: src/test/resources/hibernate.cfg.xml [hunk 4]
```diff
@@ -1,24 +1,33 @@
...
 		<!-- Select our SQL dialect -->
-		<property name="dialect">org.hibernate.dialect.MySQL5Dialect</property>
+		<!--property name="dialect">org.hibernate.dialect.MySQL5Dialect</property-->
+		<!--property name="dialect">org.hibernate.dialect.HSQLDialect</property-->
 		<!-- Echo the SQL to stdout -->
-		<property name="show_sql">true</property>
+		<property name="show_sql">false</property>
 		<!-- Set the current session context -->
 		<property name="current_session_context_class">thread</property>
 		<!-- Drop and re-create the database schema on startup -->
```
Comment (gkorland) [inline src/test/resources/hibernate.cfg.xml:29]: @MeirShpilraien the comment is wrong now

File: src/test/resources/hibernate.cfg.xml [hunk 5]
```diff
@@ -1,24 +1,33 @@
...
-		<property name="connection.username">root</property>
-		<property name="connection.password">my-secret-pw</property>
+		<!--property name="connection.driver_class">org.mariadb.jdbc.Driver</property>
+		<property name="connection.url">jdbc:mysql://localhost:3306/test?allowPublicKeyRetrieval=true&amp;useSSL=false</property-->
+		<property name="hibernate.connection.driver_class">oracle.jdbc.driver.OracleDriver</property>
+		<property name="hibernate.connection.url">jdbc:oracle:thin:@localhost:1521:xe</property>
+                <property name="hibernate.connection.username">system</property>
+                <property name="hibernate.connection.password">oracle</property>
+		<property name="hibernate.dialect">org.hibernate.dialect.OracleDialect</property>
+		<!--Property name="hbm2ddl.auto">create-drop</property-->
```
Comment (gkorland) [inline src/test/resources/hibernate.cfg.xml:15]: @MeirShpilraien remove commented out configuration

File: src/main/java/com/redislabs/WriteThroughMD.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
+package com.redislabs;
+
+import gears.GearsFuture;
+
+public class WriteThroughMD {
+  
+  private String streamID;
+  private GearsFuture<String> f;
+  private long startTime;
+  private long timeout;
```
Comment (gkorland) [inline src/main/java/com/redislabs/WriteThroughMD.java:10]:
```suggestion
  private final String streamID;
  private final GearsFuture<String> future;
  private final long startTime;
  private final long timeout;
```

File: src/main/java/com/redislabs/WriteSource.java [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+import gears.operations.GearsFutureOnDone;
+import gears.operations.OnRegisteredOperation;
+import gears.operations.OnUnregisteredOperation;
+import gears.readers.CommandOverrider;
+import gears.readers.KeysReader;
+import gears.records.KeysReaderRecord;
+import oracle.ucp.common.waitfreepool.Tuple;
+
+
+//Object pendingsWrites = GearsBuilder.execute("xlen", String.format("%s-{%s}", streamName, GearsBuilder.hashtag()));
```
Comment (gkorland) [inline src/main/java/com/redislabs/WriteSource.java:41]: Dead code?

File: src/main/java/com/redislabs/WriteSource.java [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+public class WriteSource extends Source{
+  
+  /**
+   * 
+   */
+  private static final long serialVersionUID = 1L;
+  private String streamName;
+  private int timeout;
+  
+  private boolean writeThrough;
```
Comment (gkorland) [inline src/main/java/com/redislabs/WriteSource.java:55]:
```suggestion
  private static final long serialVersionUID = 1L;
  
  private final String streamName;
  private final int timeout;
  private final boolean writeThrough;
```

**Review by gkorland — CHANGES_REQUESTED** [review]

File: src/main/java/com/redislabs/WriteThroughMD.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+  
+  public WriteThroughMD(String streamID, GearsFuture<String> f, long timeout) {
+    this.streamID = streamID;
+    this.f = f;
+    this.timeout = timeout;
+    this.startTime = System.currentTimeMillis();
+  }
+  
+  public boolean TryFree(String currStreamID) throws Exception {
```
Comment (sazzad16) [inline src/main/java/com/redislabs/WriteThroughMD.java:19]:
```suggestion
  public boolean tryFree(String currStreamID) throws Exception {
```

File: src/main/java/com/redislabs/Source.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+@JsonTypeInfo(use=JsonTypeInfo.Id.NAME, include=JsonTypeInfo.As.PROPERTY, property="type")
+public abstract class Source implements OnRegisteredOperation, OnUnregisteredOperation, Iterable<Object>{
+  
+  /**
+   * 
+   */
+  private static final long serialVersionUID = 1L;
+
+  static Map<String, Source> sources = new ConcurrentHashMap<>();
```
Comment (sazzad16) [inline src/main/java/com/redislabs/Source.java:35]:
```suggestion
  protected static final Map<String, Source> sources = new ConcurrentHashMap<>();
```

File: pytest/test_basic.py [hunk 11]
```diff
@@ -0,0 +1,508 @@
...
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        signal.setitimer(signal.ITIMER_REAL, 0)
+        signal.signal(signal.SIGALRM, signal.SIG_DFL)
+
+    def handler(self, signum, frame):
+        self.env.assertTrue(False, message=self.msg)
+        raise Exception(self.msg)
+
+def Connect():
```
Comment (sazzad16) [inline pytest/test_basic.py:57]:
```suggestion
def connect():
```

File: pytest/test_basic.py [hunk 12]
```diff
@@ -0,0 +1,508 @@
...
+    ConnectionStr = 'oracle://{user}:{password}@{db}'.format(user='system', password='oracle', db='localhost:1521/xe')
```
Comment (sazzad16) [inline pytest/test_basic.py:58]:
```suggestion
    connection_str = 'oracle://{user}:{password}@{db}'.format(user='system', password='oracle', db='localhost:1521/xe')
```

File: pytest/test_basic.py [hunk 13]
```diff
@@ -0,0 +1,508 @@
...
+    engine = create_engine(ConnectionStr).execution_options(autocommit=True)
+    conn = engine.connect()
+    return conn
+
+def GetConnection():
```
Comment (sazzad16) [inline pytest/test_basic.py:63]:
```suggestion
def get_connection():
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: pom.xml [hunk 14]
```diff
@@ -71,6 +72,8 @@
 			<groupId>com.redislabs</groupId>
 			<artifactId>gear_runtime</artifactId>
 			<version>0.0.3-SNAPSHOT</version>
+			<scope>system</scope>
+			<systemPath>${basedir}/bin/RedisGears_JVMPlugin/gears_runtime/target/gear_runtime-jar-with-dependencies.jar</systemPath>
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply pom.xml:76]: It needs to be fixed on the JVM plugin repository ...

File: pom.xml [hunk 15]
```diff
@@ -182,6 +185,7 @@
 				<artifactId>maven-assembly-plugin</artifactId>
 				<version>2.5.4</version>
 				<configuration>
+					<finalName>rghibernate</finalName>
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply pom.xml:188]: You tell me, not a mvn expert ...

File: src/main/java/com/redislabs/Connector.java [hunk 16]
```diff
@@ -0,0 +1,418 @@
...
+OnUnregisteredOperation, Iterable<String>,
+MapOperation<HashMap<String, Object>, HashMap<String, Object>>{
+  
+  public static final String ENTETY_NAME_STR = "__entityName__";
+  public static final String EVENT_STR = "__event__";
+  public static final String SOURCE_STR = "__source__";
+  
+  class MyStandardServiceInitiator implements StandardServiceInitiator<Service>{
+
+    private Map values;
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply src/main/java/com/redislabs/Connector.java:48]: It's coming from initiateService, so I thought to leave it as is.

File: src/main/java/com/redislabs/Connector.java [hunk 17]
```diff
@@ -0,0 +1,418 @@
...
+  
+  static Collection<Connector> getAllConnectors() {
+    return connectors.values();    
+  }
+  
+  static Connector GetConnector(String name) {
+    return connectors.get(name);
+  }
+  
+  private String name;
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply src/main/java/com/redislabs/Connector.java:96]: Nope, they are set on deserialization.

File: src/main/java/com/redislabs/RGHibernate.java [hunk 18]
```diff
@@ -17,38 +22,53 @@
...
+    return ret;
+  }
+  
+  public static RGHibernate get(String name) {
+    return hibernateConnections.get(name);
+  }
+  
+  private String name;
+  private String xmlConf;
+  private Map<String, String> sources;
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply src/main/java/com/redislabs/RGHibernate.java:48]: Those are set on deserialization

File: src/main/java/com/redislabs/RGHibernate.java [hunk 19]
```diff
@@ -17,38 +22,53 @@
...
   private transient Session session = null;
   private transient SessionFactory factory = null;
   private transient StandardServiceRegistry registry = null;
 
-  public RGHibernate(String[] args) {
-    this.args = args;
+  public RGHibernate(String name) {
+    this.name = name;
+    this.sources = new HashMap<>();
+    hibernateConnections.put(this.name, this);
```
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply src/main/java/com/redislabs/RGHibernate.java:58]: So where should we do it?

File: src/main/java/com/redislabs/WriteSource.java (hunk 8 above)
Comment (MeirShpilraien) [reply src/main/java/com/redislabs/WriteSource.java:55]: Set on deserialization

File: pom.xml (hunk 15 above)
Comment (sazzad16) [inline pom.xml:188]:
```suggestion
					<finalName>${jar.finalName}</finalName>
```

File: src/main/java/com/redislabs/Connector.java [hunk 20]
```diff
@@ -0,0 +1,441 @@
...
+   */
+  private static final long serialVersionUID = 1L;
+
+  public static final Map<String, Connector> connectors = new ConcurrentHashMap<>();
+  
+  static Collection<Connector> getAllConnectors() {
+    return connectors.values();    
+  }
+  
+  static Connector GetConnector(String name) {
```
Comment (sazzad16) [inline src/main/java/com/redislabs/Connector.java:91]:
```suggestion
  static Connector getConnector(String name) {
```

File: src/main/java/com/redislabs/RGHibernate.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (sazzad16) [reply src/main/java/com/redislabs/RGHibernate.java:58]: In `getOrCreate` method. Did I forgot to suggest renaming that method?

**Review by sazzad16 — CHANGES_REQUESTED** [review]

**Comment by MeirShpilraien** [comment]
- [ ] add tests for read through

File: src/main/java/com/redislabs/RGHibernate.java (hunk 19 above)
[the remark this answers is not in the document]
Comment (MeirShpilraien) [reply src/main/java/com/redislabs/RGHibernate.java:56]: renamed and moved

File: src/main/java/com/redislabs/RGHibernate.java [hunk 21]
```diff
@@ -17,38 +20,55 @@
...
-  private String[] args;
+  private static final Map<String, RGHibernate> hibernateConnections = new HashMap<>();
+  
+  public static RGHibernate getOrCreate(String name) {
+    RGHibernate ret = hibernateConnections.get(name);
+    if(ret == null) {
+      ret = new RGHibernate(name);
+    }
+    hibernateConnections.put(ret.name, ret);
+    return ret;
```
Comment (sazzad16) [inline src/main/java/com/redislabs/RGHibernate.java:38]:
```suggestion
    RGHibernate ret = hibernateConnections.get(name);
    if(ret == null) {
      ret = new RGHibernate(name);
      hibernateConnections.put(ret.name, ret);
    }
    return ret;
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

File: src/main/java/com/redislabs/RGHibernate.java (hunk 21 above)
Comment (MeirShpilraien) [reply src/main/java/com/redislabs/RGHibernate.java:38]: I want to add it only if it's new, otherwise it's already there.

Comment (sazzad16) [reply src/main/java/com/redislabs/RGHibernate.java:38]: That's exactly what I'm doing here...

Comment (MeirShpilraien) [reply src/main/java/com/redislabs/RGHibernate.java:38]: Ohh you are right, thanks.

**Review by gkorland — APPROVED** [review]

**Pull request merged**